src/Contracts/RequestPaginator.php

```diff
@@ -0,0 +1,191 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Saloon\Contracts;
+
+use Countable;
+use GuzzleHttp\Promise\PromiseInterface;
+use Iterator;
+use ReturnTypeWillChange;
+
+/**
+ * @template TRequest of \Saloon\Contracts\Request
+ * @template TResponse of \Saloon\Contracts\Response
+ *
+ * @extends \Iterator<TResponse|\GuzzleHttp\Promise\PromiseInterface>
+ */
+interface RequestPaginator extends Countable, Iterator
+{
+    /**
+     * @param callable(int $pendingRequests): (int)|int $concurrency
+     * @param callable(TResponse $response, array-key $key, \GuzzleHttp\Promise\PromiseInterface $poolAggregate): (void)|null $responseHandler
+     * @param callable(mixed $reason, array-key $key, \GuzzleHttp\Promise\PromiseInterface $poolAggregate): (void)|null $exceptionHandler
+     */
+    public function pool(
+        callable|int $concurrency = 5,
+        ?callable $responseHandler = null,
+        ?callable $exceptionHandler = null,
+    ): Pool;
+
+    /**
+     * @return $this
+     */
+    public function async(bool $async = true): static;
+
+    /**
+     * @return bool
+     */
+    public function isAsync(): bool;
+
+    /**
+     * Makes the Paginator continue where it left off in an earlier loop, instead of rewinding/'resetting' when used in a new loop.
+     *
+     * Iterables are, by nature, rewinding ('resetting') whenever you use it in a new loop.
+     * I.e., if you go through 2 iterations in a loop, then break out of it, and iterate again in a new loop, it'll be rewound to the first element.
+     * This makes sure that the rewinding step is skipped.
+     *
+     * @param bool $ignoreRewinding
+     *
+     * @return $this
+     *
+     * @TODO Come up with a better name for this method.
+     */
+    public function ignoreRewinding(bool $ignoreRewinding = true): static;
```

Comment:
I like `ignoreRewinding`, you could call it `preventRewinding`?

Reply:
Very clever way of doing it though!

Reply:
Testing `enableRewinding()` as of now. Will see how I feel about it as we go. 😄